### ISSUE TITLE
Separate C++ translation from compilation

### ISSUE THREAD

**Issue by paul-buerkner** [stan-dev/cmdstanr#276]
For running tests on the generated Stan code in brms, I would like to use the syntax checking of the compiler to at least catch any errors at compilation time. In rstan, it is possible to do translation to C++ (including syntax checking) without actual C++ compilation. It seems that cmdstanr's `$compile()` always does both at the same time. Is it possible to separate these two processes and make them available in separate functions?

**Comment by rok-cesnovar** [comment]
I think `$compile()` should do what it does now. 

Maybe just a `mod$syntax_check()`? Or is observing the C++ code required?

**Comment by paul-buerkner** [comment]
`mod$syntax_check()` would make me perfectly happy!

**Comment by rok-cesnovar** [comment]
That should be simple to add. 

Until we merge that, you can use https://rok-cesnovar.github.io/stanc3js-demo/index.html if the only thing you need is a syntax check. This uses the latest stanc3 version of the compiler and if there is an error, you will see the error, otherwise, you will see the generated cpp. It is not polished style-wise so apologies for that :)

**Comment by paul-buerkner** [comment]
This is very good news! 

Thank you also for the link. What I need in brms is programmatic syntax checking for use in unit tests, so I will probably need to wait for the feature to be available in cmdstanr.

**Comment by jgabry** [comment]
This is a good idea! I wonder if it would also eventually help with this: https://github.com/stan-dev/rstan/issues/821 (RStudio uses RStan to syntax check Stan files but RStan is way behind)

**Comment by paul-buerkner** [comment]
I am somewhat confused by the `quiet` argument of `check_syntax`. The doc says 

"quiet: (logical) Should informational messages be printed? Default is TRUE, which will print a message if the model is valid or the compiler error message if there are syntax errors in the model. If FALSE, only the error message will be printed." 

but shouldn't this be logically the other way round?

**Comment by rok-cesnovar** [comment]
Yeah, that should be 

"quiet: (logical) Should informational messages be printed? Default is FALSE, which will print a message if the model is valid or the compiler error message if there are syntax errors in the model. If TRUE, only the error message will be printed."

**Comment by paul-buerkner** [comment]
In my quick test quiet = true didnt actually silence the message when the
syntax was correct

Rok Češnovar <notifications@github.com> schrieb am Do., 27. Aug. 2020,
15:49:

**Comment by rok-cesnovar** [comment]
Hm, can you try this snippet and report back if you see any output on quiet = TRUE. It works ok for me (Ubuntu).

```
> stan_program <- file.path(cmdstan_path(), "examples", "bernoulli","bernoulli.stan")
> model <- cmdstan_model(stan_program, compile = FALSE)
> model$check_syntax(quiet = TRUE)
  
> model$check_syntax(quiet = FALSE)
Stan program is syntactically correct
```

**Comment by paul-buerkner** [comment]
You are right. Sorry, `quiet` works correctly but was just incorrectly documented. Thanks for the clarification.

**Comment by rok-cesnovar** [comment]
Fixed the docs on master. Thanks for the report!
